crates/test-utils/src/lib.rs

```diff
@@ -1,46 +1,54 @@
-use std::sync::atomic::{AtomicUsize, Ordering};
-
 use arbitrary::{Arbitrary, Unstructured};
-use rand::{rngs::OsRng, RngCore};
+use rand::{rngs::OsRng, CryptoRng, RngCore};
 
 pub mod bitcoin;
 pub mod bridge;
 pub mod evm_ee;
 pub mod l2;
 
-const ARB_GEN_LEN: usize = 1 << 24; // 16 MiB
+// Smaller buffer size as compared to 2^24
+const ARB_GEN_LEN: usize = 128;
 
 pub struct ArbitraryGenerator {
-    buf: Vec<u8>,
-    off: AtomicUsize,
+    buf: Vec<u8>, // Persistent buffer
 }
-
 impl Default for ArbitraryGenerator {
     fn default() -> Self {
         Self::new()
     }
 }
-
 impl ArbitraryGenerator {
     pub fn new() -> Self {
-        Self::new_with_size(ARB_GEN_LEN)
+        ArbitraryGenerator {
+            buf: vec![0u8; ARB_GEN_LEN],
+        }
+    }
+    pub fn new_with_size(s: usize) -> Self {
+        ArbitraryGenerator { buf: vec![0u8; s] }
     }
 
-    pub fn new_with_size(n: usize) -> Self {
-        let mut buf = vec![0; n];
-        OsRng.fill_bytes(&mut buf); // 128 wasn't enough
-        let off = AtomicUsize::new(0);
-        ArbitraryGenerator { buf, off }
+    /// Legacy interface: no arguments
+    pub fn generate<'a, T>(&'a mut self) -> T
+    where
+        T: Arbitrary<'a> + Clone,
+    {
+        self.generate_with_rng::<T, OsRng>(None)
     }
 
-    pub fn generate<'a, T: Arbitrary<'a> + Clone>(&'a self) -> T {
-        // Doing hacky atomics to make this actually be reusable, this is pretty bad.
-        let off = self.off.load(Ordering::Relaxed);
-        let mut u = Unstructured::new(&self.buf[off..]);
-        let prev_off = u.len();
-        let inst = T::arbitrary(&mut u).expect("failed to generate arbitrary instance");
-        let additional_off = prev_off - u.len();
-        self.off.store(off + additional_off, Ordering::Relaxed);
-        inst
+    /// Core function: accepts an optional RNG
+    pub fn generate_with_rng<'a, T, R>(&'a mut self, rng: Option<&mut R>) -> T
+    where
+        T: Arbitrary<'a> + Clone,
+        R: RngCore + CryptoRng,
+    {
```

Comment:
This interface is clunky.  Have a fn that take an rng unconditionally and the other fn will just call that with a default ThreadRng/OsRng/whatever.